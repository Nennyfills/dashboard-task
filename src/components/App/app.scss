.dash-app {
 .Toastify__toast-container {
  z-index: 7777777 !important;
 }
}
