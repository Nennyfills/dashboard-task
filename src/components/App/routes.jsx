import Loader from 'components/Loader';
import Records from 'pages';

const routes = [
  {
    path: '/',
    element: <div>index</div>,
    loader: <Loader remove />,
  },
  {
    path: 'block',
    element: <div>block</div>,
    loader: <Loader />,
  },
  {
    path: 'monitor',
    element: <div>monitor</div>,
    loader: <Loader />,
  },
  {
    path: 'reports',
    element: <Records />,
    loader: <Loader />,
  },
  {
    path: 'switch',
    element: <div>last route</div>,
    loader: <Loader />,
  },
];

export default routes;
