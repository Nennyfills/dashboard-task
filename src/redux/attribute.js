import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import httpWrapper from 'api/axiosWrapper';

export const getAllData = createAsyncThunk('report/getAll', async ({ url, id, filterPayload }) =>{
  const res = await httpWrapper({ url });
  if (res) {
    const calls = res?.data?.map(async ()=> await httpWrapper({
      url: 'report',
      payload: filterPayload,
      method: 'post',
    }));
    const result = await Promise.all(calls);
    const allProject = result.map((val, index)=> {
      val.name = res?.data[index]?.name;
      val.id = res?.data[index]?.[id];
      val.total = val?.data?.reduce((current, next)=> next.amount + current, 0);
      return val;
    });
    return allProject;
  }
  return [];
});

export const getReport = createAsyncThunk('report/getReport', async ({ filterPayload, title }) =>{
  const eachReport = await httpWrapper({ url: 'report', payload: filterPayload, method: 'post' });
  const data = [{
    name: title,
    data: eachReport?.data,
    total: eachReport?.data?.reduce((current, next)=> next.amount + current, 0)
  }];
  return data;
});

export const getProjects = createAsyncThunk('report/getProjects', async () =>{
  const projects = await httpWrapper({ url: 'projects' });
  return projects?.data || [];
});

export const getGateways = createAsyncThunk('report/getGateways', async () =>{
  const gateways = await httpWrapper({ url: 'gateways' });
  return gateways?.data || [];
});

export const getUser = createAsyncThunk('report/getUser', async () =>{
  const users = await httpWrapper({ url: 'users' });
  return users?.data[0];
});

export const initialState = {
  allProject: [],
  gateways: [],
  projects: [],
  user: {},
  loading: true,
};
export const attributeSlice = createSlice({
  name: 'attributes',
  initialState,
  extraReducers: {
    [getUser.fulfilled]: (state, { payload }) => {
      state.user = payload;
      state.loading = false;
    },
    [getUser.pending]: (state) => {
      state.loading = true;
    },
    [getAllData.fulfilled]: (state, { payload }) => {
      state.allProject = payload;
      state.loading = false;
    },
    [getAllData.pending]: (state) => {
      state.loading = true;
    },
    [getReport.fulfilled]: (state, { payload }) => {
      state.allProject = payload;
      state.loading = false;
    },
    [getReport.pending]: (state) => {
      state.loading = true;
    },
    [getProjects.fulfilled]: (state, { payload }) => {
      state.projects = payload;
      state.loading = false;
    },
    [getProjects.pending]: (state) => {
      state.loading = true;
    },
    [getGateways.fulfilled]: (state, { payload }) => {
      state.gateways = payload;
      state.loading = false;
    },
    [getGateways.pending]: (state) => {
      state.loading = true;
    },
  },
});

export default attributeSlice.reducer;
