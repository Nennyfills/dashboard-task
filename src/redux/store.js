import { configureStore } from '@reduxjs/toolkit';
import Attribute from './attribute';

export default configureStore({ reducer: { Attribute } });
