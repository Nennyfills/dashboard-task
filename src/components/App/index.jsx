import { Suspense, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { Route, Routes } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';

import { Layout } from 'components';
import { getUser } from 'redux/attribute';
import routes from './routes';

import 'react-toastify/dist/ReactToastify.css';

const App = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getUser());
  }, []);

  return (
    <main data-testid="app">
      <Layout>
        <ToastContainer />
        <Routes>
          {routes.map((val) => (
            <Route
              key={val?.path}
              element={<Suspense fallback={val.loader}>{val.element}</Suspense>}
              path={val?.path}
            />
          ))}
        </Routes>
      </Layout>
    </main>
  );
};

export default App;
