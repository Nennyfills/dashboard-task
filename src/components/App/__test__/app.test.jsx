import { screen } from '@testing-library/react';
import { textTender, withoutRender } from 'test/testUtils';
import configureMockStore from 'redux-mock-store';
import { attributeSlice } from 'redux/attribute';
import { configureStore } from '@reduxjs/toolkit';
import App from '..';

const mockStore = configureMockStore();
const state = mockStore({
  allProject: [],
  gateways: [],
  projects: [],
  user: {},
  loading: true,
});

const store = configureStore({
  reducer: attributeSlice.reducer,
  initialState: state,
});

const actionName = () => ({
  type: 'report/getUser/fulfilled',
  payload: {
    userId: 'rahej',
    firstName: 'John',
    lastName: 'Doe',
    email: 'john.doe@email.com',
  },
});

it('Should take a snapshot', () => {
  store.dispatch(actionName());
  const { asFragment } = textTender(<App />, store);
  expect(asFragment(withoutRender(<App />, store))).toMatchSnapshot();
});
it('Should render app without crashing', () => {
  textTender(<App />, store);
  const appElement = screen.getByTestId('app');
  expect(appElement).toBeInTheDocument();
});
