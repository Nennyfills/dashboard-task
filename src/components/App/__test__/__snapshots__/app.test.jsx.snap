// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should take a snapshot 1`] = `
<DocumentFragment>
  <main
    data-testid="app"
  >
    <section
      class="layout"
      data-testid="layout"
    >
      <header
        class="layout--navbar"
        data-testid="layout--navbar"
      >
        <nav
          class="navbar"
          data-testid="navbar"
        >
          <div
            class="navbar__icon"
            data-testid="icons"
          >
            <svg
              class="navbar__icon--logo"
            >
              logo.svg
            </svg>
            <svg
              class="navbar__icon--menu"
            >
              menuIcon.svg
            </svg>
          </div>
          <div
            class="navbar__avatar"
          >
            <div
              class="navbar__avatar__holder"
            >
              <div
                class="navbar__avatar--initials"
                data-testid="initials"
              >
                <span
                  class="h2"
                />
                <span
                  class="h2"
                />
              </div>
            </div>
            <div
              class="navbar__avatar__details"
              data-testid="details"
            >
              <p
                class="bold family roboto"
              >
                N/A
              </p>
            </div>
          </div>
        </nav>
      </header>
      <aside
        class="layout--sidebar"
        data-testid="layout--sidebar"
      >
        <section
          class="sidebar"
          data-testid="sidebar"
        >
          <section
            class="sidebar--body"
            data-testid="sidebar--body"
          >
            <a
              class=""
              data-testid="sidebar--link"
              href="/"
            >
              <svg
                class="active "
              >
                chart.svg
              </svg>
            </a>
            <a
              class=""
              data-testid="sidebar--link"
              href="/block"
            >
              <svg
                class=""
              >
                block.svg
              </svg>
            </a>
            <a
              class=""
              data-testid="sidebar--link"
              href="/monitor"
            >
              <svg
                class=""
              >
                monitor.svg
              </svg>
            </a>
            <a
              class=""
              data-testid="sidebar--link"
              href="/reports"
            >
              <svg
                class=""
              >
                report.svg
              </svg>
            </a>
            <a
              class=""
              data-testid="sidebar--link"
              href="/switch"
            >
              <svg
                class=""
              >
                switch.svg
              </svg>
            </a>
          </section>
        </section>
      </aside>
      <section
        class="layout--section"
        data-testid="layout--section"
      >
        <div
          class="layout--section--content"
          data-testid="layout--section--content"
        >
          <div
            class="Toastify"
          />
          <div>
            index
          </div>
        </div>
      </section>
      <footer
        class="layout--footer"
        data-testid="layout--footer"
      >
        <div
          class="footer-container"
          data-testid="footer-container"
        >
          <p
            class="bold"
            data-testid="footer-p"
          >
            Terms&Conditions | Privacy policy
          </p>
        </div>
      </footer>
    </section>
  </main>
</DocumentFragment>
`;
